src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -507,6 +540,64 @@ protected static DeduplicateAggregationDefinition qualify(DeduplicateAggregation
     return builder.build();
   }
 
+  protected static boolean supportsWindowAggregationDefinition(WindowAggregationDefinition def) {
+    Collection<Expression> partitionExpressions =  def.getPartitionExpressions();
+    Collection<Expression> selectExpressions = def.getSelectExpressions().values();
+    Collection<WindowAggregationDefinition.OrderByExpression> orderByExpressions = def.getOrderByExpressions();
+    Collection<Expression> orderExpressions = new ArrayList<>(orderByExpressions.size());
+    for (WindowAggregationDefinition.OrderByExpression orderByExpression : orderByExpressions) {
+      orderExpressions.add(orderByExpression.getExpression());
+    }
+    return supportsExpressions(partitionExpressions) && supportsExpressions(selectExpressions)
+      && supportsExpressions(orderExpressions);
+  }
+
+  protected static WindowAggregationDefinition qualify(WindowAggregationDefinition def) {
+    return WindowAggregationDefinition.builder().select(qualifyKeys(def.getSelectExpressions()))
+      .partition(def.getPartitionExpressions()).orderBy(def.getOrderByExpressions())
+      .windowFrameType(def.getWindowFrameType()).unboundedFollowing(def.getUnboundedFollowing())
+      .unboundedPreceding(def.getUnboundedPreceding()).following(def.getFollowing()).preceding(def.getPreceding())
+      .build();
+  }
+
+  @VisibleForTesting
+  @Nullable
+  protected static String collectWindowAggregationDefinitionErrors(WindowAggregationDefinition def) {
+    // If the expression is valid, this must be null.
+    if (supportsWindowAggregationDefinition(def)) {
+      return null;
+    }
+
+    // Gets all expressions defined in this definition
+    Collection<Expression> selectExpressions = def.getSelectExpressions().values();
+    Collection<Expression> partitionExpressions = def.getPartitionExpressions();
+    Collection<WindowAggregationDefinition.OrderByExpression> orderByExpressions = def.getOrderByExpressions();
+    Collection<Expression> orderExpressions = new ArrayList<>(orderByExpressions.size());
+    for (WindowAggregationDefinition.OrderByExpression orderByExpression : orderByExpressions) {
```

Comment:
This code is duplicated in several places. Can you make this another method?

Reply:
Done